src/main/webapp/authorization.js

```diff
@@ -1,56 +1,143 @@
+    /** @type {string} GitHub API token */
+    this.token = null;
+
+    /**
+     * the Firebase instance associated with a given
+     * GitHubAuthorizer
+     * @type {Firebase}
+     */
+    this.firebase = firebase; // eslint-disable-line
+
+    this.initializeFirebase();
+  }
+
+  /**
+   * The most recent access token for the GitHub API. To refresh this, the user
+   * must be signed out and signed in again. It is important to keep in mind
+   * that this token expires and must be used swiftly after calling
+   * toggleSignIn(). If authorization is yet to be toggled, this will return
+   * null. If an error occurs during authorization, it will again be set to
+   * null. This property will be set to null on sign out as well.
+   *
+   * NOTE: In the event that an access token times out, the token will still be
+   * available here and so, a non-null token is not a good sign of a valid
+   * token.
+   *
+   * @return {string} most recent token or null
+   */
+  getToken() {
+    return this.token;
+  }
+
+  /**
+   * The current user if authorized and null if not. For all intents and
+   * purposes, this function will serve as a true/false value in an if
+   * statement.
+   *
+   * @return {Firebase.User} current user or null
+   */
+  getUser() {
+    return this.getFirebase().auth().currentUser;
+  }
+
+  /**
+   * Get firebase instance used with this object
+   * @return {Firebase} firebase instance
+   */
+  getFirebase() {
+    return this.firebase;
+  }
+
+  /**
+   * Function called when authorizing user to verify them as a gitHub user.
+   * This function doubles as an opt out when the current user wants to leave
+   * our site for good.
+   *
+   * Side Effects: when authorizing a user, it sets the current objects token
+   * to the returned GitHub API accessToken if a token is returned.
+   *
+   * NOTE: this function is asynchronous and should be used with an await
+   *
+   * @return {AuthorizationEnum} true if no errors occured during
+   * authorization, false otherwise.
+   */
+  async toggleAuthorization() {
+    if (!this.getFirebase().auth().currentUser) {
+      try {
+        let provider = new this.firebase.auth.GithubAuthProvider();
+        let authorizationResults =
+          await this.getFirebase().auth().signInWithPopup(provider);
+        // This gives you a GitHub Access Token.
+        // You can use it to access the GitHub API.
+        this.token = await authorizationResults.credential.accessToken;
+      } catch (error) {
+        // Handle Errors here.
+        console.error(error);
+        // TODO: Change this in future iterations
+        alert(error.message);
+        this.token = null;
+        return {
+          gitHubToken: null,
+          isValid: false,
+        };
+      }
+    } else {
+      // if logged in, sign out
+      this.getFirebase().auth().signOut();
+      this.token = null;
+    }
+    return {
+      gitHubToken: this.token,
+      isValid: true,
+    };
+  }
```

Comment:
The way that this is structured is proving rather difficult to work with. Consider something like this, where each piece of functionality is individually accessible or accessible with `toggleSignIn()`:

```suggestion
  signIn() {
    let provider = new this.firebase.auth.GithubAuthProvider();
    return this.getFirebase().auth().signInWithPopup(provider)
        .then((result) => {
      this.token = result.credential.accessToken;
      return result;
    });
  }

  signOut() {
    return this.getFirebase().auth().signOut().then(() => {
      this.token = null;
      return null;
    });
  }

  toggleSignIn() {
    if (this.getUser()) {
      return this.signOut();
    } else {
      return this.signIn();
    }
  }
```

Reply:
This also allows the user of the class to define how errors are handled.

Reply:
yeah that's pretty good